### PR TITLE
Cleanup tests properly to avoid session leaks

### LINKED ISSUE #310: Make sure integration tests close spanner objects
Should call connection.close() after #284 is done.

But also integration tests use spanner client library objects directly ([example](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/blob/9ae2b756586157cc43b3d6b865e6a341f56b7150/cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java#L560), so tests should call close methods on the client object.

This issue is brought to you bu a failing `getSessionNames()` test that could no longer find currently active session in the first 1000 entry page of results.

### PR DESCRIPTION
**Pull request by elefeint**
Noticed during fixit that we were running out of available server side sessions. 

Spanner client library-based ConnectionFactory now implements R2DBC `Closeable`, as discussed in #284. Closing that out requires a small bit of documentation that I will do separately.

Fixes #310.

### CODE REVIEW

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java [hunk 1]
```diff
@@ -127,6 +129,11 @@ public void deleteData() {
         .block();
   }
 
+  @AfterAll
+  static void cleanUpEnvironment() {
```
Comment (dzou) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java:133]: Remove? (if not used)

File: cloud-spanner-r2dbc-samples/cloud-spanner-r2dbc-sample/src/main/java/com/example/BookExampleApp.java [hunk 2]
```diff
@@ -50,14 +54,19 @@ public BookExampleApp(String sampleInstance, String sampleDatabase,
         .option(INSTANCE, sampleInstance)
         .option(DATABASE, sampleDatabase)
         .build());
+
+    this.connection = Mono.from(this.connectionFactory.create()).block();
+  }
+
+  public void cleanup() {
+    this.connection.close();
```
Comment (dzou) [inline cloud-spanner-r2dbc-samples/cloud-spanner-r2dbc-sample/src/main/java/com/example/BookExampleApp.java:62]: Do you need to block on this? `this.connection.close().block()`

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+/**
+ * This class does not follow test naming conventions intentionally.
+ * It is intended for one-off use troubleshooting and cleaning up sessions.
+ */
+public class SessionCleanupUtils {
```
Comment (dzou) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java:43]: This is not meant to get picked up by Maven right? Maybe add a `@Ignore` to be explicit.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  static SpannerStub spannerStub;
+
+  @BeforeAll
+  static void setUpEnvironment() throws Exception {
+    client = new GrpcClient(GoogleCredentials.getApplicationDefault());
+    spannerStub = client.getSpanner();
+  }
+
+  @Test
+  void deleteAllSessions() throws Exception {
```
Comment (meltsufin) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java:55]: Should the Spanner client library have this functionality? It seems odd that we have to implement cleanup logic for sessions.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryConnectionFactory.java [hunk 5]
```diff
@@ -55,4 +56,15 @@ public SpannerClientLibraryConnectionFactory(SpannerConnectionConfiguration conf
...
   }
+
+  /**
+   * This method is blocking, to be called at the end of the application lifecycle.
+   * @return
+   */
+  @Override
+  public Publisher<Void> close() {
+    // TODO: check if it would still be blocking in Cloud Spanner Connection API
+    this.spannerClient.close();
```
[the remark this answers is not in the document]
Comment (elefeint) [reply cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryConnectionFactory.java:67]: yep.

File: cloud-spanner-r2dbc-samples/cloud-spanner-r2dbc-sample/src/main/java/com/example/BookExampleApp.java (hunk 2 above)
Comment (elefeint) [reply cloud-spanner-r2dbc-samples/cloud-spanner-r2dbc-sample/src/main/java/com/example/BookExampleApp.java:62]: yes!

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java (hunk 1 above)
Comment (elefeint) [reply cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java:133]: It's used; I will comment it back out. I was experimenting with leaving dirty sessions to test the session removal script.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java (hunk 4 above)
Comment (elefeint) [reply cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SessionCleanupUtils.java:55]: During normal operation of the client library, there should be no leftover sessions left. In R2DBC driver v1, leftover sessions were due to tests not calling `Connection.close()`. In v2, they were due to `ConnectionFactory.close()` not yet being implemented. I am adding a comment to `SessionCleanupUtils` that during normal operation, no leftover sessions are expected.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java (hunk 1 above)
Comment (elefeint) [reply cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/ClientLibraryBasedIT.java:133]: aaaand commenting it out in the first place was the cause of my leftover session leaks. Truly, I am both the solution to and the cause of all of my problems.

**Review by dzou — APPROVED** [review]

**Review by meltsufin — APPROVED** [review]
Note the 5 minor code smells, but otherwise LGTM.

**Pull request merged**